### PR TITLE
[V1][Spec Decode] Fix MTP bugs and enable MLA support

### PR DESCRIPTION
**Pull request by benchislett**
## Overview

This PR enables Flashinfer-MLA to work with MTP in vLLM V1. This is accomplished by padding the inputs in `prepare_inputs_deferred` for the EAGLE drafter. See #21984 for details.

This draft PR includes a number of other changes:
- Changes from #21078 are included in this PR as a dependency. It should be merged first.
- Performance improvements which allow the MLA+MTP pathway to synchronize cpu/gpu only after the speculative forward pass.
- Multiple bugs are preventing MTP from working properly, some are fixed by this PR. In my testing I have identified:
- - Issues in weight loading when using `VLLM_DISABLE_MLA`, which is not fixed by this PR
- - Bug causing failure running MTP due to using kwargs explicitly in `eagle.py` when calling the model forward, which is fixed by this PR
- - Inconsistent support for `reorder_batch_threshold`, which is refactored by this PR
- - Bug causing failure when running EAGLE3 + Llama 3.1 8B with cuda graphs enabled on Blackwell, not fixed by this PR (#22755)

## Review Notes

I wish to guide reviewers towards some specific topics of discussion related to the design decisions made in this implementation:

### Enablement options
In this PR, each attention backend must opt-in to being able to handle qlen > 1 by overriding `decode_supports_qlen_padding`.  Also, the Attention Metadata Builders must implement `get_reorder_batch_threshold` if they will support decode query lengths greater than 1. This is currently done at the MLAAttentionMetadataBuilder level, which will cause failures for MLA backends other than FlashInferMLA backend when trying to use this feature. 

Problematically, the MLAAttentionMetadataBuilder cannot see the AttentionBackend class instance and query `decode_supports_qlen_padding` to decide how to set `reorder_batch_threshold` effectively. I am sure there is a clean way to implement these two flags in order to enable/disable this support on a per-backend basis. Please advise.

### Performance refactors and correctness 
Some changes from #20078 are adapted for this PR to enable the 'deferred' (aka 'padded') pathway to synchronize the cpu/gpu only after the EAGLE forward pass(es), leading to a theoretical performance improvement.

These changes are comprehensive and involve moving a lot of calculation from the CPU to the GPU. In the future these could be adapted into kernels (see #20078 for an example) to further reduce overhead. 

I have evaluated the correctness of this feature for K=1,2, and 3 speculative tokens, BS > 1, with the FlashInfer-MLA backend on DeepSeek-R1 on 8xB200. However, there are currently multiple critical bugs stopping me from evaluating the performance compared to a baseline (see above), so the exact performance advantages of this approach cannot be fully evaluated yet.

### CODE REVIEW

File: vllm/model_executor/models/deepseek_mtp.py [hunk 1]
```diff
@@ -158,14 +158,13 @@ def forward(
         self,
         input_ids: torch.Tensor,
         positions: torch.Tensor,
-        previous_hidden_states: torch.Tensor,
+        hidden_states: torch.Tensor,
```
Comment (benchislett) [inline vllm/model_executor/models/deepseek_mtp.py:161]: NOTE: This fixes a critical bug breaking MTP support, since the arguments are now passed as `kwargs` by `eagle.py` and therefore must be called `hidden_states`.

File: vllm/v1/attention/backends/mla/flashmla.py [hunk 2]
```diff
@@ -161,8 +161,18 @@ def _forward_decode(
...
-        q = torch.cat([q_nope, q_pe], dim=-1)\
-            .unsqueeze(1) # Add seqlen dim of 1 (decode)
+        q = torch.cat([q_nope, q_pe], dim=-1)
+
+        batch_size = attn_metadata.decode.block_table.shape[0]
+
+        needs_padding = q.shape[0] % batch_size != 0
+
+        if needs_padding:
+            raise ValueError("oops")
```
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/v1/attention/backends/mla/flashmla.py:171]: This is temp development code, I have not been able to confirm this works since I am developing primarily on Hopper. The final flashinfer-mla design can be replicated to flashmla once it is polished.

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -1787,25 +1825,53 @@ def propose_draft_token_ids(
...
-                    next_token_id = req_state.get_token_id(seq_len)
-                next_token_ids.append(next_token_id)
-            next_token_ids = torch.tensor(next_token_ids,
-                                          dtype=torch.int32,
-                                          device=self.device)
+            assert discard_sampled_tokens_req_indices is not None
+
+            _max_gen_len = sampled_token_ids.shape[-1]
+            # Get all sampled tokens from valid requests
+            _valid_sampled_token_ids_gpu = sampled_token_ids
```
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/v1/worker/gpu_model_runner.py:1834]: I am concerned about the perf impacts here, and I have verified that the write is safe as this function always returns a new sampled_token_ids output to take the place of the input one

**Comment by benchislett** [comment]
Closing for now, see #24539 for follow-up
